### ISSUE TITLE
'CS8629: Nullable value type may be null' reported when it can not be null after HasValue

### ISSUE THREAD

**Issue by davidhunter22** [dotnet/roslyn#36175]
This occurs with C# 8

```
public struct Foo
{
    int get_value( )
   {
        int? value = get_value_from_somewhere( );

        // This test will not generate a 'Nullable value type may be null' 
        // if( value == null ) return 0;

        // This test will not generate a 'Nullable value type may be null' 
        // if( !value.HasValue ) return 0;

        // This test will generate a 'Nullable value type may be null' 
        if( value.HasValue == false ) return 0;

        return value.Value; // Warning on this line
    }

    // In reality do something that may or may not return a value
    int? get_value_from_somewhere( ) => 0; 
}
```
I think the last test is simple enough that it should not generate a warning

**Comment by jcouv** [comment]
This is an issue with the `== false` throwing off the nullability analysis. I agree this seems simple enough that we should be able to handle it.

Similarly, `== true` causes an unexpected warning in `if(value.HasValue == true) return value.Value;`

**Comment by jcouv** [comment]
This is tracked by https://github.com/dotnet/roslyn/issues/29855 already.
I'll go ahead and close this as a duplicate. Thanks for reporting this!

**Comment by ezavaleta** [comment]
LINQ queries raise this warning too.

 ```csharp
var list = new int? [] { 0, 1, 2, 3, null, 5 };
var not_null = list.Where (x => x.HasValue).Select (x => x.Value).ToList (); // Warning CS8629: Nullable value type may be null.
var not_null2 = list.Where (x => x != null).Select (x => x.Value).ToList (); // Warning CS8629: Nullable value type may be null.
 ```
